tests/telemetry_test.py

```diff
@@ -295,6 +295,142 @@ def test_sets_options_for_java_11_or_above_custom_recording_template(self):
+            "-XX:StartFlightRecording=maxsize=0,maxage=0s,disk=true,dumponexit=true,"
+            "filename=/var/log/test-recording.jfr,delay=10s,duration=20m",
+        ]
+
+    def test_sets_options_for_java_11_or_above_custom_delay_and_duration(self):
+        jfr = telemetry.FlightRecorder(
+            telemetry_params={"delay": "10s", "duration": "20m"}, log_root="/var/log", java_major_version=random.randint(11, 999)
+        )
+        assert jfr.java_opts("/var/log/test-recording.jfr") == [
+            "-XX:+UnlockDiagnosticVMOptions",
+            "-XX:+DebugNonSafepoints",
+            "-XX:StartFlightRecording=maxsize=0,maxage=0s,disk=true,dumponexit=true,"
+            "filename=/var/log/test-recording.jfr,delay=10s,duration=20m",
+        ]
+
+    def test_sets_options_for_pre_java_9_custom_delay_duration_recording_template(self):
+        jfr = telemetry.FlightRecorder(
+            telemetry_params={"recording-template": "profile", "duration": "20m", "delay": "10s"},
+            log_root="/var/log",
+            java_major_version=random.randint(0, 8),
+        )
+        assert jfr.java_opts("/var/log/test-recording.jfr") == [
+            "-XX:+UnlockDiagnosticVMOptions",
+            "-XX:+DebugNonSafepoints",
+            "-XX:+UnlockCommercialFeatures",
+            "-XX:+FlightRecorder",
+            "-XX:FlightRecorderOptions=disk=true,maxage=0s,maxsize=0,dumponexit=true,dumponexitpath=/var/log/test-recording.jfr",
+            "-XX:StartFlightRecording=defaultrecording=true,delay=10s,duration=20m,settings=profile",
+        ]
+
+    def test_sets_options_for_java_9_or_10_custom_delay_duration_recording_template(self):
+        jfr = telemetry.FlightRecorder(
+            telemetry_params={"recording-template": "profile", "duration": "20m", "delay": "10s"},
+            log_root="/var/log",
+            java_major_version=random.randint(9, 10),
+        )
+        assert jfr.java_opts("/var/log/test-recording.jfr") == [
+            "-XX:+UnlockDiagnosticVMOptions",
+            "-XX:+DebugNonSafepoints",
+            "-XX:+UnlockCommercialFeatures",
+            "-XX:StartFlightRecording=maxsize=0,maxage=0s,disk=true,dumponexit=true,"
+            "filename=/var/log/test-recording.jfr,delay=10s,duration=20m,settings=profile",
+        ]
+
+    def test_sets_options_for_java_11_or_above_custom_delay_duration_recording_template(self):
+        jfr = telemetry.FlightRecorder(
+            telemetry_params={"recording-template": "profile", "duration": "20m", "delay": "10s"},
+            log_root="/var/log",
+            java_major_version=random.randint(11, 999),
+        )
+        assert jfr.java_opts("/var/log/test-recording.jfr") == [
+            "-XX:+UnlockDiagnosticVMOptions",
+            "-XX:+DebugNonSafepoints",
+            "-XX:StartFlightRecording=maxsize=0,maxage=0s,disk=true,dumponexit=true,"
+            "filename=/var/log/test-recording.jfr,delay=10s,duration=20m,settings=profile",
+        ]
+
```

Comment:
Thanks for being that thorough in your tests. However, I think it's a bit too much. Each test should test for a specific thing, and each change should only break one test. Case in point, to change duration to jfr-duration and delay to jfr-delay, you now have to change twelve tests!

In practice, you don't have to test specifically for the Java version, it's done already. And you don't have to test for the "no parameters case", it's done already. You can reduce your twelve tests to one with all the settings enabled (params, delay and duration).

Reply:
Makes sense, will be updating this to just have a single test will all params.